lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -826,6 +826,36 @@ LogicalResult BitsPrimOp::canonicalize(BitsPrimOp op,
                                             newLo);
     return success();
   }
+  // Mux narrowing
+  // bits(mux(x,a,b)) -> mux(x,bits(a),bits(b))
+  if (auto innerMux = dyn_cast_or_null<MuxPrimOp>(inputOp)) {
+    auto muxTypeWidth = innerMux.getResult()
+                            .getType()
+                            .cast<FIRRTLType>()
+                            .getBitWidthOrSentinel();
+    if (muxTypeWidth > 0 && innerMux->hasOneUse() == 1) {
+      // FIRRTL lets constants to muxes be narrower than the mux, which means we
+      // may need to extend the arguments when we move the bitsPrimOp.
+      auto newHi = innerMux.high();
+      if (newHi.getType().cast<FIRRTLType>().getBitWidthOrSentinel() <
+          muxTypeWidth)
```

Comment:
This will trigger when the operand has unknown width, which is probably the wrong thing.  I'd recommend wrapping this whole pattern in a check to `hasKnownWidthIntTypesAndNonZeroResult`.